desc/singularities.py

```diff
@@ -0,0 +1,733 @@
+    # but then need to convert vectors back to rpz
+    if kernel.ndim == 3:
+        f = xyz2rpz_vec(f, phi=eval_data["zeta"])
+    return f
+
+
+def _singular_part(
+    eval_data,
+    eval_grid,
+    src_data,
+    src_grid,
+    s,
+    q,
+    kernel,
+    interpolator,
+    loop=False,
+):
+    """Integrate singular point by interpolating to polar grid."""
+    src_dtheta = src_grid.spacing[:, 1]
+    src_dzeta = src_grid.spacing[:, 2] / src_grid.NFP
+    eval_theta = jnp.asarray(eval_grid.nodes[:, 1])
+    eval_zeta = jnp.asarray(eval_grid.nodes[:, 2])
+
+    r, w, dr, dw = _get_quadrature_nodes(q)
+
+    h_t = jnp.mean(src_dtheta)
+    h_z = jnp.mean(src_dzeta)
+
+    eta = _chi(r)
+    v = eta * s**2 * h_t * h_z / 4 * abs(r) * dr * dw
+    keys = list(set(["|e_theta x e_zeta|"] + kernel.keys))
+    fsrc = [src_data[key] for key in keys]
+
+    def polar_pt_vmap(i):
+        # evaluate the effect from a single polar node around each singular point
+        # on that singular point. Polar grids from other singularities have no effect
+        dt = s / 2 * h_t * r[i] * jnp.sin(w[i])
+        dz = s / 2 * h_z * r[i] * jnp.cos(w[i])
+        theta_i = eval_theta + dt
+        zeta_i = eval_zeta + dz
+
+        # data interpolated to each eval pt offset by dt,dz
+        src_data_polar = {key: interpolator(val, i) for key, val in zip(keys, fsrc)}
+
+        src_data_polar["zeta"] = zeta_i
+        src_data_polar["theta"] = theta_i
+
+        # eval pts x src pts for 1 polar grid offset
+        # only need diagonal term because polar grid points
+        # don't contribute to other eval pts
+        k = kernel(
+            {key: val for key, val in eval_data.items() if key in keys},
+            src_data_polar,
+            diag=True,
+        ).reshape((eval_grid.num_nodes, kernel.ndim))
+
+        dS = (v[i] * src_data_polar["|e_theta x e_zeta|"])[:, None]
+        fi = k * dS
+        return fi
+
+    def polar_pt_loop(i, f):
+        # this calculates the effect at a single evaluation point, from all others
+        # in a single field period. loop this to get all pts
+        f_temp = polar_pt_vmap(i)
+        return f + f_temp.reshape((eval_grid.num_nodes, kernel.ndim))
+
+    f = jnp.zeros((eval_grid.num_nodes, kernel.ndim))
+    # vmap found more efficient than fori_loop, esp on gpu, but uses more memory
+    if loop:
+        f = fori_loop(0, v.size, polar_pt_loop, f)
+    else:
+        f = vmap(polar_pt_vmap)(jnp.arange(v.size)).sum(axis=0)
+
+    # we sum distance vectors, so they need to be in xyz for that to work
+    # but then need to convert vectors back to rpz
+    if kernel.ndim == 3:
+        f = xyz2rpz_vec(f, phi=eval_data["zeta"])
+
+    return f
+
+
+def singular_integral(
+    eval_data,
+    eval_grid,
+    src_data,
+    src_grid,
+    kernel,
+    interpolator,
+    loop=False,
+):
+    """Evaluate a singular integral transform on a surface.
+
+    eg f(θ, ζ) = ∫ ∫ K(θ, ζ, θ', ζ') g(θ', ζ') dθ' dζ'
```

Comment:
I know this is in the paper but can you also define here to specify that K is the kernel and `g` is the metric (right?)

Reply:
Or can you cite the equation? (this one seems like eqn 3.7 in [1], though without the $\sigma` function)

this would be helpful down the line for when we are deceased and others are maintaining the code in a 100 years.